### PR TITLE
Support for repeating qualifiers in ArC plus automated tests

### LINKED ISSUE #7732: Support repeatable qualifier annotations
https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#repeating_qualifiers

### PR DESCRIPTION
**Pull request by manovotn**
Fixes #7732

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java [hunk 1]
```diff
@@ -51,11 +51,12 @@ static BeanInfo createClassBean(ClassInfo beanClass, BeanDeployment beanDeployme
...
         for (AnnotationInstance annotation : beanDeployment.getAnnotations(beanClass)) {
-            if (beanDeployment.getQualifier(annotation.name()) != null) {
+            for (AnnotationInstance qualifierAnnotation : beanDeployment.getQualifierExtractor(annotation.name())
+                    .apply(annotation)) {
                 // Qualifiers
-                qualifiers.add(annotation);
-                if (DotNames.NAMED.equals(annotation.name())) {
-                    AnnotationValue nameValue = annotation.value();
+                qualifiers.add(qualifierAnnotation);
+                if (DotNames.NAMED.equals(qualifierAnnotation.name())) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java]: I don't think we need to run this for every annotation. I'd place this `if`  block outside the loop.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java [hunk 2]
```diff
@@ -51,11 +51,12 @@ static BeanInfo createClassBean(ClassInfo beanClass, BeanDeployment beanDeployme
         String name = null;
 
         for (AnnotationInstance annotation : beanDeployment.getAnnotations(beanClass)) {
-            if (beanDeployment.getQualifier(annotation.name()) != null) {
+            for (AnnotationInstance qualifierAnnotation : beanDeployment.getQualifierExtractor(annotation.name())
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java]: Hm, maybe we could simplify the code a little bit. `Function<> getQualifierExtractor(annotation.name())` -> `Collection<AnnotationInstance> getQualifiers(annotation)` and the loop could be replaced with `beanDeployment.getQualifiers(annotation).forEach(qualifiers::add);`. WDYT?

**Review by mkouba — CHANGES_REQUESTED** [review]
I think that `Beans` is not the only place where we should handle repeatable qualifiers. Although the spec only mentions the type safe resolution it should work for observers, disposers, injection points, etc. In other words, we should find all occurences of `io.quarkus.arc.processor.BeanDeployment.getQualifier(DotName)` and consider which should be replaced.

Otherwise looks really good!

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java (hunk 2 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java]: I just realized that my code doesn't call `continue` properly on the outer loop once we know it's qualifier (so we needn't process it any further). And so does yours. We should take that into account, so in fact it should look something like this:

```
            Collection<AnnotationInstance> qualifierCollection = beanDeployment.getQualifierExtractor(annotation.name()).apply(annotation);
            for (AnnotationInstance qualifierAnnotation : qualifierCollection) {
                // Qualifiers
                qualifiers.add(qualifierAnnotation);
            }
            if (!qualifierCollection.isEmpty()) {
                // we needn't process it further, the annotation was a qualifier (or multiple repeating ones)
                continue;
            }
```

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java [hunk 3]
```diff
@@ -363,6 +372,20 @@ ClassInfo getQualifier(DotName name) {
         return qualifiers.get(name);
     }
 
+    Function<AnnotationInstance, Collection<AnnotationInstance>> getQualifierExtractor(DotName name) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: Looking at the usage I wonder if it makes sense to return a function. `Collection<AnnotationInstance> getQualifiers(AnnotationInstance annotation)` would be sufficient and more convenient in most cases. Alternatively, we could add a convenient method like this:
```java
Collection<AnnotationInstance> getQualifiers(AnnotationInstance annotation) {
    return getQualifierExtractor(annotation.name()).apply(annotation);
}
```

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java [hunk 4]
```diff
@@ -95,6 +97,12 @@
     private final boolean removeFinalForProxyableMethods;
     private final boolean jtaCapabilities;
 
+    private Function<AnnotationInstance, Collection<AnnotationInstance>> singletonFunction = annotationInstance -> Collections
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: These could be final... but it's just a minor.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java [hunk 5]
```diff
@@ -71,9 +72,11 @@ void init(List<Throwable> errors) {
     }
 
     Collection<AnnotationInstance> getDisposedParameterQualifiers() {
-        return Annotations.getParameterAnnotations(declaringBean.getDeployment(), disposerMethod, disposedParameter.position())
-                .stream().filter(a -> declaringBean.getDeployment().getQualifier(a.name()) != null)
-                .collect(Collectors.toList());
+        Set<AnnotationInstance> resultingQualifiers = new HashSet<>();
+        Annotations.getParameterAnnotations(declaringBean.getDeployment(), disposerMethod, disposedParameter.position())
+                .stream().forEach(a -> declaringBean.getDeployment().getQualifierExtractor(a.name()).apply(a)
+                        .forEach(an -> resultingQualifiers.add(an)));
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java]: `resultingQualifiers::add`?

**Review by mkouba — CHANGES_REQUESTED** [review]

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java (hunk 3 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: Ok, I will change that to the suggested method. You're right that it's easier to read and use.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java (hunk 4 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: These will be removed altogether accordingly to your earlier suggestion.

**Comment by manovotn** [comment]
Fixed the things you asked me for - the util method is longer operates on functions.
However, I also found a glitch in test I wrote and fixing it showed me an error. I am now looking at whether that is just me being silly on Mon or whether we have a bug in disposer method resolution.

**Comment by manovotn** [comment]
Turns out it was an actual bug - in typesafe resolution done for disposer methods with qualifiers.
I will fix it as part of this PR.

**Comment by manovotn** [comment]
Updated, I am once again getting green builds locally :)

**Review by mkouba — APPROVED** [review]

**Pull request merged**